src/iarc7_motion/high_level_motion_controller.py

```diff
@@ -0,0 +1,213 @@
+
+import iarc_tasks.task_states as task_states
+import iarc_tasks.task_commands as task_commands
+
+
+class HighLevelMotionController:
+
+    def __init__(self, action_server):
+        # action server for getting requests from AI
+        self._action_server = action_server
+        
+        # current state of HLM
+        self._task = None
+        self._initialized = False
+
+        # used for timeouts & extrapolating 
+        self._timer = None
+        self._timeout_vel_sent = False 
+
+        # to keep things thread safe
+        self._lock = threading.RLock()
+
+        # handles monitoring of state of drone
+        self._state_monitor = StateMonitor()
+
+        # handles communicating between tasks and LLM
+        self._task_command_handler = TaskCommandHandler()
+
+        # safety 
+        self._safety_client = SafetyClient('high_level_motion')
+        self._safety_land_complete = False
+        self._safety_land_requested = False
+
+        # Create action client to request a safety landing
+        self._action_client = actionlib.SimpleActionClient(
+                                        "motion_planner_server",
+                                        QuadMoveAction)
+        try:
+            # update rate for HLM
+            self._update_rate = rospy.get_param('~update_rate')
+            # task timeout values
+            self._task_timeout = rospy.Duration(rospy.get_param('~task_timeout'))
+
+        except KeyError as e:
+            rospy.logerr('Could not lookup a parameter for High Level Motion Controller')
+            raise
+
+    def run(self):
+        # rate limiting of updates of HLM
+        rate = rospy.Rate(self._update_rate)
+
+        rospy.logwarn('trying to form bond')
+
+        # forming bond with safety client
+        if not self._safety_client.form_bond():
+            rospy.logerr('High Level Motion could not form bond with safety client')
+            return
+
+        rospy.logwarn('done forming bond')
+
+        self._action_client.wait_for_server()
+
+        while not rospy.is_shutdown():
+            # Exit immediately if fatal
+            if self._safety_client.is_fatal_active() or self._safety_land_complete:
+                return
+
+            # Land if put into safety mode
+            if self._safety_client.is_safety_active() and not self._safety_land_requested:
+                # Request landing
+                goal = QuadMoveGoal(movement_type="land", preempt=True)
+                self._action_client.send_goal(goal,
+                        done_cb=self._safety_task_complete_callback)
+                rospy.logwarn('motion planner attempting to execute safety land')
+                self._safety_land_requested = True
+
+            # if we have not seen a task yet
+            # we wait until now to start the task timeout timer
+            # because the action clients have not yet been initialized
+            if not self._initialized:
```

Comment:
Line 148 is for when we just had a task and it just ended. 

We only want to create new timers once and not every time, so as soon as a task is set to none, you only create a timer once. The `initialized` is there on line 103 because 1) we need to wait until we are actually in the main loop and 2) we only want to create a timer once, as things could get messy creating new instances all the time (as you can also see in the need for a careful shutdown of a timer each time we need to).